suppress management command

jingcheng16 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Related to [SUPPORT-17293](https://dimagi-dev.atlassian.net/browse/SUPPORT-17293?atlOrigin=eyJpIjoiZTM4Njc1MzlhMmQ5NDdkNGE3NzJjNDk2MGVmMGUwMDQiLCJwIjoiaiJ9)

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Get all the duplicate invoices, and suppress the one with a later `date_created` 
Checked how our suppress worked, realized it's just setting `is_hidden_to_ops` to `True`.
Then for all the invoices, if the `subtotal` and `balance` is not equal, it means certain amount of payment have already being made, so we have to credit them back.

Credit will be given at subscription level not account level. For any credit that have a type: plan, sms, or user, we credit them back to the CreditLine have the same type.

Note: The last commit 2a555b4 is for reproducing error on staging. This commit will be reverted once I successfully reproduce the error.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
No

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally
- If we have duplicated invoices, this will suppress the duplicate ones
- If not, the output of the command would be: `Successfully suppressed 0 duplicate invoices for date: 2023-07-01`

Haven't fully test for credit them back, because I don't know how to make a "fake" payment other than credit...

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
I didn't write any test for this command. Should I?
Considering the time sensitivity of this ticket, I expect to run this command ASAP... so I'm a bit rushed in getting this PR out.


### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Maybe no QA. Since the duplicate invoices didn't happen in staging environment.


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SUPPORT-17293]: https://dimagi-dev.atlassian.net/browse/SUPPORT-17293?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jingcheng16 committed:
suppress management command

dannyroberts commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,36 @@
+                subscription_id=sub_id, date_start=invoice_start_date).order_by('-date_created')
+            for invoice in related_invoices[1:]:
+                invoice.is_hidden_to_ops = True
+                invoice.save()
+                suppressed_count += 1
```
In scripts like this is a good idea to provide detailed output of what actions were performed, which you can then copy and save as a record on the ticket.

So here that might be

```suggestion
                print(f"Suppressing invoice {invoice.id} ({invoice.subscription.subscriber.domain} ", end='')
                invoice.save()
                suppressed_count += 1
                print(f"✓")
```
or something like that.

jingcheng16 replied:
Thanks @dannyroberts ! Fixed here: 2aa271213171a8527169e10a9661ecb1a87139c9

dannyroberts reviewed (commented):
Is the plan to do credit adjustments in a separate command?

millerdev reviewed (dismissed):
I don't think any of my feedback is blocking. The first question is mostly for my own understanding.

jingcheng16 committed:
provide detailed output

jingcheng16 commented:
> Is the plan to do credit adjustments in a separate command?

@dannyroberts Yes, I need more time to study how credit adjustments works. I think the sooner we suppress the invoice, we can avoid more clients double paying us.

jingcheng16 commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,36 @@
+        invoice_start_date = kwargs['invoice_start_date']
+
+        # Filter the Invoice objects based on the date_start
+        invoices_of_given_date = Invoice.objects.filter(date_start=invoice_start_date)
```
Yes, they all have the same `date_start`, so for invoices in July, the `date_start` will always be `2023-07-01`.

jingcheng16 replied:
@millerdev Thank you for asking this important question.
At first I think, yes, all invoices generated for the previous month will start at the first day of the previous month, since I remember [this line of code ](https://github.com/dimagi/commcare-hq/blob/e9b562403bf45ea5b2e0081561722a433e6b0f73/corehq/apps/accounting/tasks.py#L288)in our periodic `generate_invoices` function. 
But your question just keep running in my background, and I just realized what if a client starting at the middle of the month???
Then I find out, the `date_start` can be any date in a given month...
I'll repull a list of data and also make changes to my script tomorrow.

jingcheng16 commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,36 @@
+                subscription_id=sub_id, date_start=invoice_start_date).order_by('-date_created')
+            for invoice in related_invoices[1:]:
+                invoice.is_hidden_to_ops = True
+                invoice.save()
```
Yes! Fixed here: https://github.com/dimagi/commcare-hq/commit/2aa271213171a8527169e10a9661ecb1a87139c9

jingcheng16 committed:
simplify the query

jingcheng16 commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,36 @@
+            count=Count('id')).filter(count__gt=1)
+
+        # Extract the subscription ids from the above queryset
+        duplicate_subs_ids = [item['subscription'] for item in duplicate_subscriptions]
```
Thanks! Fixed here: bdf40c39b29b5a67df07984a147d15e5288211b3

millerdev commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,37 @@
+        duplicate_subs_ids = invoices_of_given_date.values('subscription').annotate(
+            count=Count('id')).filter(count__gt=1).values_list('subscription', flat=True)
+
+
```
nit: lint
```suggestion
```

jingcheng16 committed:
lint

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

jingcheng16 committed:
Credit them back

jingcheng16 committed:
Suppress the later one

jingcheng16 committed:
exclude 0 credit

jingcheng16 committed:
applied_credit is negative

jingcheng16 commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,84 @@
+            )
+            if feature_deduction:
+                print(f"Adding {feature[0]} credit: {feature_deduction}")
+                CreditLine.add_credit(amount=feature_deduction, subscription=invoice.subscription,
```
Known issue here: `feature[0]` would give me "User" and "SMS", however the `feature_type` in CreditLine model should be "USER" and "SMS"... I don't know how to do that, and I don't want to manually capitalize the word... Can I get some advice on this?

AmitPhulera replied:
> however the feature_type in CreditLine model should be "USER" and "SMS"

This was a bit of a surprise to me given that we have `FeatureType` class with static values. To verify this I queried production to see the values and I found out only `User` and `SMS` are used. 

Shared the queries on Slack.

I think you should be good with using `User` and `SMS` here. Let me know if this answers your questions or if I misunderstood it.

jingcheng16 replied:
Thank you! When I was playing around with CreditLine, I tried to generate each type of CreditLine in shell. So the “USER” record I saw in my local environment, is actually my own “wrong setup Credit Line”, not something we originally have in database. Will mark as resolved.

AmitPhulera commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,84 @@
+                                    feature_type=feature[0], note=note)
+                payment_by_credit -= feature_deduction
+
+        # After deducting plan credit and feature predit, the remained credit should be type Any
```
```suggestion
        # After deducting plan credit and feature credit, the remained credit should be type Any
```

AmitPhulera commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,84 @@
+            'Successfully suppressed {} duplicate invoices for Statement Period: {}-{}'
+            .format(suppressed_count, year, month)))
+
+    def revert_invoice_payment(self, invoice, note):
```
For my understanding - How did we finalize on the implementation on reverting the invoice payment? Is this documented somewhere? Maybe in a ticket or some place else?

jingcheng16 replied:
Documented [here](https://docs.google.com/document/d/1TY2Io4Y7r7IGIHdShg--sHgiWavB35jkjo3VDsiO1M8/edit)

AmitPhulera commented:
@jingcheng16  Overall changes looks good. No blockers from my side. I have a question regarding implementation of `revert_invoice_payments`. 

And one question I have - Do we have plans to test this on staging before running it on production?

jingcheng16 committed:
typo

Co-authored-by: Amit Phulera <amitphulera999@gmail.com>

jingcheng16 commented:
> And one question I have - Do we have plans to test this on staging before running it on production?

@AmitPhulera 
Not yet.
I can try generate duplicate invoices, by removing our constraints for generating duplicate invoices and create a private release, then call the function from my private release. But I don't know how to fake some payments on staging... Ask QA now, will update once I get response.

jingcheng16 committed:
generate duplicate for staging on purpose

AmitPhulera commented:
Thanks for trying. Issues like these sometimes get too tricky to reproduce. I think its fine to move ahead. You can remove your test commit.

jingcheng16 committed:
Revert "generate duplicate for staging on purpose"

This reverts commit 2a555b4bbaebf2a5ec5a191ab0842f0ca502bcba.

jingcheng16 commented:
> Thanks for trying. Issues like these sometimes get too tricky to reproduce. I think its fine to move ahead. You can remove your test commit.

@AmitPhulera Thank you.  Reverted.

jingcheng16 committed:
lint

millerdev reviewed (commented):
I don't have enough experience with the process of crediting invoices to feel confident signing off on this.

biyeun commented:
@jingcheng16 Thanks for this! The code looks generally good, but without tests I'm a bit hesitant to fully sign off. 
In place of tests, I would also be okay with a "dry run" version of the command where all the actions are printed before fully executed. You can see examples of this with other management commands if you search for `dry_run`.

biyeun commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,85 @@
+        start_date, end_date = get_first_last_days(year, month)
+
+        # Filter the Invoice objects based on the date_start
+        invoices_of_given_date = Invoice.objects.filter(date_start__range=(start_date, end_date))
```
we should also filter by `is_hidden_to_ops=False`

jingcheng16 replied:
You're right! Fixed here: ecac1a31cf8492dcaddaad85f38583a1aea04df5

biyeun reviewed (changes requested):
I found a blocking change that's definitely needed
line 25 most definitely needs an additional filter `is_hidden_to_ops=False`

jingcheng16 committed:
Do not suppress invisible invoice

jingcheng16 committed:
include reference to the first created invoice

jingcheng16 committed:
Add dry run version

jingcheng16 committed:
lint

dannyroberts reviewed (commented):
This looks good to me. +1 to Biyeun's dry-run suggestion; that is a very good pattern for commands like this one.

@biyeun I'll let you do the final pass since you're already engaged on it

dannyroberts commented:
> Should I do a private release and dry run it on production?
This is typically allowed be done after approval, and yeah I think it will be a good step. (We prefer not to run scripts in even in dry mode without review/approval, since part of review is reviewing that the dry run is in fact dry).

biyeun commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,97 @@
+        suppressed_count = 0
+        for sub_id in duplicate_subs_ids:
+            related_invoices = list(Invoice.objects.filter(
+                subscription_id=sub_id, date_start__range=(start_date, end_date)).order_by('-date_created'))
```
I think we also need the filter `is_hidden_to_ops=False` here as well

jingcheng16 replied:
Ah yes! Fixed here: ff3d7a571ad7ebb54d6d35185ed7c19734137dda

biyeun commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,97 @@
+        month = kwargs['month']
+        note = kwargs['note']
+        dry_run = kwargs['dry_run']
+        dry_run_tag = '[DRY_RUN]' if dry_run else ''
```
```suggestion
        dry_run_tag = '[DRY RUN] ' if dry_run else ''
```
space for readability considering how it's being used below

jingcheng16 replied:
Add spaces for both `dry_run_tag` here: d4dd616d3399a1650af84d4a51670fa0e64a4031

biyeun commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -0,0 +1,97 @@
+                                             f'duplicate invoices for Statement Period: {year}-{month}'))
+
+    def revert_invoice_payment(self, invoice, note, dry_run=False):
+        dry_run_tag = '[DRY_RUN]' if dry_run else ''
```
```suggestion
        dry_run_tag = '[DRY RUN] ' if dry_run else ''
```

biyeun commented:
thanks @jingcheng16. changes look good, just one more filter

jingcheng16 committed:
Add filter

jingcheng16 committed:
Add space for readability

biyeun reviewed (approved):
thank you @jingcheng16 looks great! 🎉

jingcheng16 commented:
@biyeun Thanks Biyeun. Should I do a private release on production and dry run it now?

AmitPhulera commented:
@jingcheng16 Since the PR is approved you can make a private release with this branch and do the dry run there and share your findings with the team.

jingcheng16 commented:
@biyeun 
This is the only invoice use SMS Credit: 
https://www.commcarehq.org/a/onusida-1/settings/project/billing/statements/download/2cc72ee494779ff0596a856c1e3a1a24.pdf

But after dry run, the SMS credit isn't "reverted". I took a look and find this weird invoice. The total should be a sum up of all cost. But it seems like it only includes plan cost.

If you go here to see all invoices for this domain:
https://www.commcarehq.org/hq/accounting/invoices/?subscriber=onusida-1&payment_status=&report_filter_statement_period_month=7&report_filter_statement_period_year=2023&report_filter_due_date_month=8&report_filter_due_date_year=2023&is_hidden=

It seems like, the total column is sometimes 0, sometimes have a value...even if we always have a plan cost.

jingcheng16 committed:
correct calculation

invoice.applied_credit doesn't include feature credit; when we add credit back, it should be a positive number, so add a minus sign before the negative credit we get

jingcheng16 committed:
fix plan credit and feature credit

jingcheng16 committed:
remove the if clause for tell if the payment is made

subtotal doesn't include those plan credit or feature credit

jingcheng16 committed:
Change log text

jingcheng16 committed:
improve log message

dannyroberts reviewed (approved):
Offlined with @jingcheng16 on the dry run output

jingcheng16 committed:
Do not return plan credit for auto generated credit

jingcheng16 commented:
@biyeun 
Made change in 555e93b due to issues described [here](https://dimagi-dev.atlassian.net/browse/SAAS-14813?focusedCommentId=282382)
Except for don't revert plan credits for those subscription, we still have to deal with those double created credit line... my plan is we check if there is two "ANY" type credit line associate with this subscription, if yes, then we delete one credit line, and the associated credit adjustment history. What do you think?

jingcheng16 committed:
Remove duplicate credit line

biyeun commented on corehq/apps/accounting/management/commands/suppress_duplicate_invoices.py:
```diff
@@ -70,6 +70,20 @@ def revert_invoice_payment(self, invoice, note, dry_run=False):
+            except CreditLine.MultipleObjectsReturned:
+                print(f"Multiple plan credit lines found for subscription {invoice.subscription.id}.")
+                duplicate_cl = CreditLine.objects.filter(account=invoice.subscription.account,
+                                                         subscription=invoice.subscription, isProduct=True,
```
just to be sure these are only the `CreditLine`s created by a previous run of the script, you should filter by `last_modified=datetime.datetime(2023, 8, 15)` or something like that :)

jingcheng16 replied:
Hi Biyeun, previous run of the script didn't create CreditLine. It's created when we generate invoice. I'll filter by `date_created=datetime.datetime(2023,08,01)`

biyeun reviewed (approved):
new changes looking good, just one comment about filtering

jingcheng16 committed:
Add date_created filter

jingcheng16 committed:
typo

jingcheng16 committed:
is_product not isProduct

jingcheng16 committed:
Fix wrong filter

jingcheng16 committed:
Add dry run tag

jingcheng16 committed:
Fix hardcoded date

jingcheng16 committed:
Generate report

jingcheng16 commented:
Made some improvements, so this command will now generate a csv file, list every duplicate invoices, with additional column like 'Will Suppress', 'Will Revert', 'Suppression Status', 'Revert Status'
This will make review easier for Ops, compared to a dry run log.
